Governance: complete proposal instruction

ochaloup opened the pull request:
#### Summary

Make it possible to transition off-chain/manually executable `Proposal` from `Succeeded` to `Completed` state.

SebastianBor commented on governance/program/tests/process_complete_proposal.rs:
```diff
@@ -0,0 +1,74 @@
+        .await
+        .unwrap();
+
+    let mut proposal_cookie = governance_test
```
nit: There is a helper function `with_signed_off_proposal` which combines `Create`, `AddSignatory` and `SingOff` instructions

SebastianBor commented on governance/program/tests/process_complete_proposal.rs:
```diff
@@ -0,0 +1,74 @@
+use spl_governance::state::enums::ProposalState;
+
+#[tokio::test]
+async fn test_execute_complete_proposal_without_transaction() {
```
nit: Since it's the basic use case for the instruction it should be named `test_complete_proposal`. This way it tells you which test to troubleshoot first in case of any problems with the instruction

SebastianBor commented on governance/program/src/processor/process_complete_proposal.rs:
```diff
@@ -0,0 +1,54 @@
+        proposal_data.closed_at = Some(clock.unix_timestamp);
+        proposal_data.state = ProposalState::Completed;
+    }
+
```
nit: It's not going to change semantics in the current implementation but let's throw an error if the `Proposal` can't be completed (plus test)

SebastianBor commented on governance/program/tests/program_test/mod.rs:
```diff
@@ -3143,4 +3143,42 @@ impl GovernanceProgramTest {
     }
+
+    #[allow(dead_code)]
+    pub async fn with_complete_proposal(
```
nit: The function doesn't return any cookie (it's imperative and doesn't create any account state) and hence should be named `complete_proposal`

SebastianBor commented on governance/program/tests/program_test/mod.rs:
```diff
@@ -3143,4 +3143,42 @@ impl GovernanceProgramTest {
+    }
+
+    #[allow(dead_code)]
+    pub async fn with_complete_proposal_using_instruction_impl<F: Fn(&mut Instruction)>(
```
nit: There is no need for the `_impl` suffix which is used for internal implementations

SebastianBor commented on governance/program/tests/program_test/mod.rs:
```diff
@@ -3143,4 +3143,42 @@ impl GovernanceProgramTest {
+            .process_transaction(&[complete_proposal_ix], None)
+            .await?;
+
+        let proposal_data = self.get_proposal_account(&proposal_cookie.address).await;
```
Do not use actual state in the test sdk functions. The idea is that they should return what is expected and then you can compare, and assert, with the actual state in the test

SebastianBor reviewed (commented):
Mostly looks good and I only added nits regarding naming and semantic conventions used in the project.

Let's also scope this to just completed `Succeeded` proposals and hence also fix the state transition in `FinaliseVote`

ochaloup commented:
@SebastianBor thank you for the points, I hope I addressed them (expecting the result code could be squashed by github here thus I added a separate commit into this PR)

ochaloup commented on governance/program/src/processor/process_complete_proposal.rs:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    // proposal execution still finishing with errors and time elapsed -> Completed
+    if proposal_data.state == ProposalState::ExecutingWithErrors
```
I would create a separate PR when this one is completed.

SebastianBor reviewed (approved):
Looks good

joncinque commented on governance/program/src/instruction.rs:
```diff
@@ -1660,3 +1666,21 @@ pub fn refund_proposal_deposit(
 }
+
+/// Complete proposal when proposal is stuck in Succeeded
+#[allow(clippy::too_many_arguments)]
```
nit: this can be removed, 2 arguments is definitely not too many :smile:

joncinque commented on governance/program/src/instruction.rs:
```diff
@@ -523,6 +523,12 @@ pub enum GovernanceInstruction {
     ///   2. `[writable]` Proposal deposit payer (beneficiary) account
     RefundProposalDeposit {},
+
+    /// Verify the state and move the Proposal to Completed.
+    /// Permission-less operation whoever could call.
+    ///
+    ///   0. `[writable]` Proposal account.
+    CompleteProposal {},
```
Nit: can you add more comments about when this is used? For example, specifying "this instruction only works for `Succeeded` proposals that have no instructions".

And perhaps the name can be changed to `CompleteEmptySuccessfulProposal` or some other name that describes the intention more.

You're probably more aware of the terms used in the space, so I'll defer to you!

ochaloup replied:
@joncinque thanks. I elaborated a bit more on function of the instruction in comment.
I tend to use the generic `CompleteProposal` name but I would like to ask @SebastianBor  for opinion on this.
This instruction should  (or could) have two parts. This PR resolves only one of them. I created a separate issue to make it clear (my wrong not doing so before).
The one that this PR resolves is
https://github.com/solana-labs/solana-program-library/issues/3772
The other (a potentially future PR) is:
https://github.com/solana-labs/solana-program-library/issues/4047

joncinque replied:
Ah that's great, thanks for clarifying!  Since all of the other comments are in the enum description, could you move your comments on `complete_proposal` up to here? And if it were me, I'd rename it to `CompleteStuckProposal`, but I'll let you figure that out :smile:

ochaloup replied:
Thank you @joncinque .

I fixed the place where the documentation comment resides at + I renamed the instruction to `CompleteStuckProposal` as you suggested.
If you find time to check it would be great.

joncinque reviewed (commented):
The logic looks great, just a question about the terms for this

joncinque reviewed (approved):
Looks good to me! I'll let @SebastianBor merge it if it looks good to him

SebastianBor commented:
After further review of the logic I've made changes to the semantics  of  `CompleteProposal` instruction because it's not abnormal situation (it's not stuck) to have an instruction-less and executable Proposal in the `Succeeded` state.

It's a valid scenario where we have a binding proposal (as opposed to opinion only) but the actions of the proposal are either off-chain or have to be executed manually. In such a scenario `CompleteProposal` allows the owner of the `Proposal` to flag it as `Completed` once the external actions are taken. 

@joncinque please give it another look given the changes in the semantics
